### ISSUE TITLE
[Bug] False negative : no type check for instance attribute declared in constructor

### ISSUE THREAD

**Issue by ErwanDL** [python/mypy#8553]
I am unsure whether this behaviour is known or not, but mypy doesn't seem to correctly infer/check the types of instance attributes that are declared in the constructor and have an implicit typing.

Example : 
```python
class Article:
    def __init__(self):
        self.title = "A title"    # self.title is implicitly of type "str"

def my_func(a: Article) -> None:
    print(a)

art = Article()

my_func("this is a string so mypy is unhappy")    # mypy(error) : Argument 1 to "my_func" has incompatible type "str"; expected "Article"
my_func(art.title)    # no error here, although there should be the same error as above
```

Note that if I explicitly annotate the type of `self.title` in the constructor for Article by one of the following means, the error correctly appears : 
```python
# By declaring Article as a dataclass
@dataclass
class Article
    title: str = "A title"

# By explicitly typing "self.title"
class Article:
    def __init__(self):
        self.title: str = "A title"

# By assigning "self.title" from a typed parameter to the constructor
class Article:
    def __init__(self, title: str):
        self.title = title

# In those 3 cases, the error correctly appears
my_func(art.title)    # mypy(error) : Argument 1 to "my_func" has incompatible type "str"; expected "Article"
```

I would be happy to try and solve this issue myself if you estimate it is not a hard fix.

**Comment by ErwanDL** [comment]
This appears to be a duplicate of https://github.com/python/mypy/issues/8493. Closing this issue, as the bug is being discussed there.
